agent/lib/kontena/service_pods/lifecycle_hook_manager.rb

```diff
@@ -0,0 +1,109 @@
+module Kontena
+  module ServicePods
+    class LifecycleHookManager
+      include Kontena::Helpers::RpcHelper
+      include Common
+
+      attr_reader :service_pod
+
+      # @param [ServicePod] service_pod
+      def initialize(service_pod)
+        @service_pod = service_pod
+      end
+
+      # @return [Boolean]
+      def on_pre_start
+        pre_start_hooks = hooks_for('pre_start')
+        return if pre_start_hooks.size == 0
+
+        pre_start_hooks.each do |hook|
+          begin
+            info "running pre_start hook: #{hook['cmd']}"
+            service_config = config_container(service_pod.dup, hook['cmd'])
+            service_container = create_container(service_config)
+            service_container.tap(&:start).attach
```

Comment:
Testing with hooks that execute very quickly (just `echo ...`), I somehow managed to get this stuck here:

```
W, [2017-08-24T11:32:27.447220 #1]  WARN -- Kontena::Agent: Thread 1wvzk9pw 
W, [2017-08-24T11:32:27.447410 #1]  WARN -- Kontena::Agent: /usr/lib/ruby/gems/2.4.0/gems/excon-0.58.0/lib/excon/socket.rb:263:in `select'
/usr/lib/ruby/gems/2.4.0/gems/excon-0.58.0/lib/excon/socket.rb:263:in `select_with_timeout'
/usr/lib/ruby/gems/2.4.0/gems/excon-0.58.0/lib/excon/socket.rb:175:in `rescue in read_nonblock'
/usr/lib/ruby/gems/2.4.0/gems/excon-0.58.0/lib/excon/socket.rb:156:in `read_nonblock'
/usr/lib/ruby/gems/2.4.0/gems/excon-0.58.0/lib/excon/socket.rb:36:in `read'
/usr/lib/ruby/gems/2.4.0/gems/excon-0.58.0/lib/excon/response.rb:163:in `parse'
/usr/lib/ruby/gems/2.4.0/gems/excon-0.58.0/lib/excon/middlewares/response_parser.rb:7:in `response_call'
/usr/lib/ruby/gems/2.4.0/gems/docker-api-1.33.6/lib/excon/middlewares/hijack.rb:45:in `response_call'
/usr/lib/ruby/gems/2.4.0/gems/excon-0.58.0/lib/excon/connection.rb:389:in `response'
/usr/lib/ruby/gems/2.4.0/gems/excon-0.58.0/lib/excon/connection.rb:253:in `request'
/usr/lib/ruby/gems/2.4.0/gems/docker-api-1.33.6/lib/docker/connection.rb:40:in `request'
/usr/lib/ruby/gems/2.4.0/gems/docker-api-1.33.6/lib/docker/connection.rb:65:in `block (2 levels) in <class:Connection>'
/usr/lib/ruby/gems/2.4.0/gems/docker-api-1.33.6/lib/docker/container.rb:123:in `attach'
/app/lib/kontena/service_pods/lifecycle_hook_manager.rb:24:in `block in on_pre_start'
/app/lib/kontena/service_pods/lifecycle_hook_manager.rb:19:in `each'
/app/lib/kontena/service_pods/lifecycle_hook_manager.rb:19:in `on_pre_start'
/app/lib/kontena/service_pods/creator.rb:50:in `perform'
/app/lib/kontena/workers/service_pod_worker.rb:120:in `ensure_running'
/app/lib/kontena/workers/service_pod_worker.rb:100:in `ensure_desired_state'
/app/lib/kontena/workers/service_pod_worker.rb:76:in `block in apply'
/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/task.rb:97:in `exclusive'
/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid.rb:421:in `exclusive'
/app/lib/kontena/workers/service_pod_worker.rb:74:in `apply'
/app/lib/kontena/workers/service_pod_worker.rb:33:in `update'
/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `public_send'
/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/calls.rb:28:in `dispatch'
/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/call/async.rb:7:in `dispatch'
/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:50:in `block in dispatch'
/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/cell.rb:76:in `block in task'
/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/actor.rb:339:in `block in task'
/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/task.rb:44:in `block in initialize'
/usr/lib/ruby/gems/2.4.0/gems/celluloid-0.17.3/lib/celluloid/task/fibered.rb:14:in `block in create'
```

`docker inspect` shows that the hook container srated and finished in ~60ms with an exit code of 0. Is there some kind of race between the `start` and `attach`?